### DISCUSSION TITLE
Infer query options for `useQueries` when using array function (`map`) as input

### DISCUSSION THREAD

**Discussion by bennettdams** [TanStack/query#3425] [Q&A] [answered]
Is there a good way to utilize type-safe query options when using an array as input for `useQueries`?

Here's a basic way to use `useQueries`:

```ts
  const result1 = useQueries(
    userIds.map((userId) => {
      return {
        queryKey: ["user", userId],
        queryFn: () => queryFn(userId),
        refetchInterval: 123
      };
    })
  );
```
In this case, `refetchInterval` is not typed correctly. That's because TypeScript's `map` function does not infer the types correctly.

You can circumvent that by giving `map` a generic for your query options:

```ts
  const result2 = useQueries(
    userIds.map<UseQueryOptions<QueryData>>((userId) => {
      return {
        queryKey: ["user", userId],
        queryFn: () => queryFn(userId),
        refetchInterval: 123
      };
    })
  );
```
Using the generic, the options are typed correctly.

I created a small CodeSandbox that shows this behavior. I deliberately left `refetchInterval` empty, so you can see the missing/given type when hovering:

https://codesandbox.io/s/priceless-matsumoto-0yiwc0?file=/src/App.tsx

Result 1 (without generic, missing type for options):

![image](https://user-images.githubusercontent.com/29319414/159341799-d2dc6bf6-bd73-46ed-9ae7-65e44afa1219.png)

Result 2 (with generic, correct type for options):

![image](https://user-images.githubusercontent.com/29319414/159341891-1f5daf9c-1aeb-413f-99d1-82630cf42dcd.png)

---

The [example for dynamic parallel queries via `useQueries` in the docs uses `map`](https://react-query.tanstack.com/guides/parallel-queries#dynamic-parallel-queries-with-usequeries), so I was wondering if we should recommend given it the correct type via the generic? 

![image](https://user-images.githubusercontent.com/29319414/159354179-528c1c65-f38b-4808-ac64-5f5813237bae.png)

...or maybe `useQueries` types can be optimized? Here's a third example which could be the optimal solution. It just takes a generic for the query data as a generic for `useQueries`. But as you can see in the CodeSandbox, this does also not work right now (query options are not typed correctly). 

```ts
  const result3 = useQueries<QueryData[]>(
    userIds.map((userId) => {
      return {
        queryKey: ["user", userId],
        queryFn: () => queryFn(userId),
        refetchInterval: 123
      };
    })
  );
```

**Comment by TkDodo** [comment, marked as answer]
despite the codesandbox, I'm not quite sure what the problem is. `refetchInterval` can be a `number`, so `refetchInterval: 123` is correct, and you'll get no type error here. If I do:

```
  const result1 = useQueries(
    userIds.map((userId) => {
      return {
        queryKey: ["user", userId],
        queryFn: () => queryFn(userId),
        refetchInterval: "f"
      };
    })
  );
```

now that type errors with:

```
Type 'string' is not assignable to type 'number | false | ((data: unknown, query: Query<unknown, unknown, unknown, QueryKey>) => number | false) | undefined'.ts(2345)
```

which also looks correct. Just because you hover something and it shows `any` doesn't mean there is an issue with the types. Please always run `tsc` to see if there are actual errors, or just something not ideal with autocompletion in the IDE.

**Reply by bennettdams** [reply]
Oh thanks, didn't think about that.

**Reply by bennettdams** [reply]
I now realized that I was confused by something else: When specifying options for `useQueries`, using  `number` for `refetchInterval` works great for inferring, but using `false` raises a type error:


https://codesandbox.io/s/eloquent-wing-8m0qiq?file=/src/App.tsx:341-566

- React Query `v3.34.16`
- TypeScript `v4.4.2`

```ts
  // Variant 1: `refetchInterval: 123`
  // no errors
  const resultTypeOk = useQueries(
    userIds.map((userId) => {
      return {
        queryKey: ["user", userId],
        queryFn: () => queryFn(userId),
        // `number`
        refetchInterval: 123
      };
    })
  );

  // Variant 2: `refetchInterval: false`
  // type error
  const resultTypeError = useQueries(
    userIds.map((userId) => {
      return {
        queryKey: ["user", userId],
        queryFn: () => queryFn(userId),
        refetchInterval: false
      };
    })
  );
```
Variant 2 says: 
```
...
Type 'boolean' is not assignable to type 'number | false | ((data: unknown, query: Query<unknown, unknown, unknown, QueryKey>) => number | false) | undefined'.ts(2345)
```

Is this related? Should I create an issue for that?

--- 
Update:
I circumvented this problem by using `as const` for the returned object, as this will restrict the type of `refetchInterval` from `boolean` to `false`, which is one of the expected types of a query option:

```ts
      return {
        queryKey: ["user", userId],
        queryFn: () => queryFn(userId),
        refetchInterval: false
      } as const;
```
Still not sure if this intended to be happening though.

**Reply by TkDodo** [reply]
yeah I think its just unfortunate that typescript infers the type `boolean` when you write `false` on an object:

```
const obj = {
    refetchInterval: false
}

type X = typeof obj
```

the `type X` will be `{ refetchInterval: boolean }`. [playground](https://www.typescriptlang.org/play?ssl=5&ssc=20&pln=1&pc=1#code/MYewdgzgLgBCBGArGBeGBvAUDHMBOApgGYFTAAWAkmFAXgG4CGANgFwxEsQGYC+mmKAE8ADgRgANVDGFiQROEiA).

`boolean` is not allowed because `true` doesn't make any sense. what you can also do is write `refetchInterval: 0` or `refetchInterval: undefined` to turn it off, or leave it out because no interval is the default. `refetchInterval: () => false` seems to also work better. `as const` works because TS will infer the most narrow type possible.

**Reply by bennettdams** [reply]
Thanks!

**Reply by reuelsilva** [reply]
Hi bennettdams.  Why we use an arrow function in queryFn: () => queryFn(userId) and it doesn't work queryFn: queryFn(userId)?
